### ISSUE TITLE
Doctrine is Executing DELETE outside transaction

### ISSUE THREAD

**Issue by trsteel88** [doctrine/orm#5799]
I have an issue where Doctrine is executing a DELETE query outside a transaction. I put together a test in this repo: https://github.com/trsteel88/symfony-standard/tree/doctrine-manytomany
1. Run app/console doctrine:schema:update
2. Go to /test/setup (this creates a product entity with 2 categories).

After running /test/setup you will be redirected to /test/run (id is the product we just created.)

On the /test/run, we are looping over all the categories and removing the product. We **aren't** flushing the EntityManager here. You are then redirected to /test/check.

On /test/check we are counting the categories. They are now empty. How is this possible. We never flushed the EntityManager.

Below is a screenshot of the DELETE query running.

![screen shot 2016-04-28 at 5 09 26 pm](https://cloud.githubusercontent.com/assets/869933/14878000/fd41ce30-0d63-11e6-973f-ea0ecf3d37bd.png)

**Comment by bendavies** [comment]
`PersistentCollection` does the deleting:

https://github.com/doctrine/doctrine2/blob/59a041095183f188faac6bc784074f4f88475a4a/lib/Doctrine/ORM/PersistentCollection.php#L376

https://github.com/doctrine/doctrine2/blob/59a041095183f188faac6bc784074f4f88475a4a/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php#L218

**Comment by trsteel88** [comment]
Hmmm, this seems like a pretty big flaw to me. Consider the following:

You load a form and untick all the product categories. Validation fails and says you need to select at least 1 category. You close the form. The categories were deleted even though you didn't flush. 

Shouldn't a transaction be started before any entity deletions can occur?

**Comment by Ocramius** [comment]
Transactions are usually implicitly execution via auto-commit in most engines. This is a normal behavior of `EXTRA_LAZY` collections: you should wrap the entire execution around the code that does DB-level writes in a transaction.

The ORM does not know when and how many operations will be done on the collection, so it isn't able to start/commit a transaction for you here.

**Comment by trsteel88** [comment]
Hmm.. this is huge.

The main reason we want to use EXTRA_LAZY is so that when we do entity.children.count() it doesn't load the entire collection. We want it to execute a count query.

Is there something inbetween EXTRA_LAZY and LAZY that allows this?

**Comment by Ocramius** [comment]
Nope. EXTRA_LAZY is where it's at. I suggest using
`EntityManager#transactional()` around your code, when possible.

**Comment by trsteel88** [comment]
Hmm, that's going to make my controllers a bit thicker.

Not sure about the internals, but could Doctrine just start a transaction prior to a remove() if one didn't exist? Then, if the EntityManager didn't flush, that delete wouldn't execute?

**Comment by Ocramius** [comment]
No, that would:

1. Be a BC break
2. Confusing as heck
3. Still force the user to manually commit

**Comment by trsteel88** [comment]
Thanks for the info @Ocramius.

To confirm, does this only happen on a ManyToMany or will this occur on a OneToMany as well.

**Comment by Ocramius** [comment]
IIRC, it may happen with orphan removal and `OneToMany`, but I'd have to check that with a test :-\

**Comment by trsteel88** [comment]
Hmm, it looks like OneToMany with an OrphanRemoval does delete it without a flush.

I never new about this. I'm surprised this isn't a bigger issue for people. Do most developers wrap all their controllers with transactions?

**Comment by Ocramius** [comment]
> Do most developers wrap all their controllers with transactions?

No, but most apps have relatively few problems with being non-transactionally safe (domains of relative transactional integrity interest).

I was teaching wrapping in transactions pretty much everything today, and it's quite easy to do so with a tiny command bus, [for example](https://github.com/Ocramius/doctrine-tutorial-bootstrapper/blob/c60e73a8a65dc8f459f32d29362389ed7beaf251/command-bus.php).

**Comment by JCMais** [comment]
> I was teaching [...]

Is the presentation available somewhere?

**Comment by Ocramius** [comment]
@JCMais no, we actually live-coded that bit, sorry. I'll come up with a presentation about it, but the file I linked is really, really simple.

**Comment by ThomasLabstep** [comment]
> I'm surprised this isn't a bigger issue for people.

It's 2019 and I just ran into this issue right now.
We have a onFlush listener to remove the child from Elasticsearch.
As we added EXTRA_LAZY to have an optimized count query, we bump into this issue.

> Do most developers wrap all their controllers with transactions?

Guess I will just add a transaction in my APIController remove child method.
Hopefully that will work.

Please document this kind of dangerous situation in the EXTRA_LAZY chapter of doctrine.

**Comment by marc-romera** [comment]
Hi, I just found this and became really worried and started looking for workarounds, but then I checked the documentation and I noticed from version 2.7.1 this is no longer happening.

Now `removeElement()` causes a full collection load, but does not perform write operations "on the fly". ([See docs](https://www.doctrine-project.org/projects/doctrine-orm/en/2.8/tutorials/extra-lazy-associations.html))

I think it's worth mentioning it in this thread in case someone ends up here like me.

I would actually consider this solved.
